packages/gatsby-source-contentful/src/rich-text.js

```diff
@@ -1,184 +1,69 @@
-  ].indexOf(node.nodeType) >= 0
-
-const isAssetReferenceNode = node =>
-  [BLOCKS.EMBEDDED_ASSET, INLINES.ASSET_HYPERLINK].indexOf(node.nodeType) >= 0
-
-const isEntryReferenceField = field => _.get(field, `sys.type`) === `Entry`
-const isAssetReferenceField = field => _.get(field, `sys.type`) === `Asset`
-
-const getFieldProps = (contentType, fieldName) =>
-  contentType.fields.find(({ id }) => id === fieldName)
-
-const getAssetWithFieldLocalesResolved = ({ asset, getField }) => {
-  return {
-    ...asset,
-    fields: _.mapValues(asset.fields, getField),
-  }
-}
-
-const getFieldWithLocaleResolved = ({
-  field,
-  contentTypesById,
-  getField,
-  defaultLocale,
-  resolvedEntryIDs,
-}) => {
-  // If the field is itself a reference to another entry, recursively resolve
-  // that entry's field locales too.
-  if (isEntryReferenceField(field)) {
-    if (resolvedEntryIDs.has(field.sys.id)) {
-      return field
+import { documentToReactComponents } from "@contentful/rich-text-react-renderer"
+import resolveResponse from "contentful-resolve-response"
+
+// We need the original id from contentful as id attribute
+// This code can be simplified as soon we properly replicate
+// Contentful's sys object
+const prepareIdsForResolving = obj => {
+  for (let k in obj) {
+    const value = obj[k]
+    if (
+      value &&
+      value.sys &&
+      value.sys.type === `Link` &&
+      value.sys.contentful_id
+    ) {
+      value.sys.id = value.sys.contentful_id
+      delete value.sys.contentful_id
+    } else if (value && typeof value === `object`) {
+      prepareIdsForResolving(value)
     }
-
-    return getEntryWithFieldLocalesResolved({
-      entry: field,
-      contentTypesById,
-      getField,
-      defaultLocale,
-      resolvedEntryIDs: resolvedEntryIDs.add(field.sys.id),
-    })
   }
-
-  if (isAssetReferenceField(field)) {
-    return getAssetWithFieldLocalesResolved({
-      asset: field,
-      getField,
-    })
-  }
-
-  if (Array.isArray(field)) {
-    return field.map(fieldItem =>
-      getFieldWithLocaleResolved({
-        field: fieldItem,
-        contentTypesById,
-        getField,
-        defaultLocale,
-        resolvedEntryIDs,
-      })
-    )
-  }
-
-  return field
 }
 
-const getEntryWithFieldLocalesResolved = ({
-  entry,
-  contentTypesById,
-  getField,
-  defaultLocale,
-
-  /**
-   * Keep track of entries we've already resolved, in case two or more entries
-   * have circular references (so as to prevent an infinite loop).
-   */
-  resolvedEntryIDs = new Set(),
-}) => {
-  const contentType = contentTypesById.get(entry.sys.contentType.sys.id)
-
-  return {
-    ...entry,
-    fields: _.mapValues(entry.fields, (field, fieldName) => {
-      const fieldProps = getFieldProps(contentType, fieldName)
-
-      const fieldValue = fieldProps.localized
-        ? getField(field)
-        : field[defaultLocale]
+function renderRichText({ raw, references }, options = {}) {
+  const richText = JSON.parse(raw)
```

Comment:
Is it possible for this `raw` data model to have circular references? If so, this call will throw an error. Please make sure to check for this.

Reply:
nope it can not by definition of a Contentful API response :)